### PR TITLE
Fix WebGPU resource cleanup on device destruction

### LINKED ISSUE #9393: WebGPU: WebgpuGraphicsDevice.destroy() never flushes _deferredDestroys nor calls GPUDevice.destroy(), leaking every resource of a destroyed device
### Description

`WebgpuGraphicsDevice.destroy()` never releases the GPU resources of the device it is destroying:

1. `WebgpuTexture.destroy()` / `WebgpuBuffer.destroy()` (and everything built on them: render targets, the MSAA/depth back buffer, `dynamicBuffers`, query sets) don't destroy the `GPUTexture` / `GPUBuffer` directly — they push it to `device.deferDestroy()`, which is only flushed at the end of `submit()`.
2. `WebgpuGraphicsDevice.destroy()` (and `destroyDeviceResources()` on `main`) destroys the clear/mipmap/resolve renderers, the quad buffers, `dynamicBuffers`, `gpuProfiler` and `backBuffer` — all of which go through `deferDestroy()` — and then returns. **Nothing flushes `_deferredDestroys` afterwards**, because no further `submit()` ever happens on a destroyed device.
3. `GPUDevice.destroy()` (`this.wgpu.destroy()`) is never called either (on `main` it is only called from `debugLoseContext()`).

So after `app.destroy()` every WebGPU resource the device owned stays alive until the browser garbage-collects the `GPUDevice` wrapper. On WebKit (WKWebView / Safari) the Metal allocations behind those resources are kept for a long time — in our editor, which creates and destroys a device every time a preview is started/stopped, this shows up as an unbounded growth of the `graphics` region of the WebContent process.

`loseContext()` on `main` already contains the exact loop that is missing from `destroy()`:

```js
for (const resource of this._deferredDestroys) {
    resource.destroy();
}
this._deferredDestroys.length = 0;
```

### Measurements

Katari (a Tauri 2 / WKWebView editor, macOS 26, Apple Silicon) with PlayCanvas 2.22.0. Each start/stop of a preview creates 2 devices and destroys 2. `footprint <WebContent pid>` before / after N round-trips:

| round-trips | WebContent `graphics` | phys footprint | engine `device._vram` (live device) |
|---|---|---|---|
| 0 | 372 MB | 822 MB | 34.7 MB |
| 5 | 880 MB | 1,315 MB | 34.7 MB |
| 10 | 1,434 MB | 1,843 MB | 34.7 MB |
| 15 | **1,961 MB** | **2,364 MB** | 34.7 MB |

Nothing is reclaimed after 60 s. A page reload brings `graphics` back to 81 MB, so the page (the dead `GPUDevice`s) is holding it. +106 MB per round-trip is consistent with the resources of the two destroyed devices (2× 2048² font atlas + a 1280×720 4× MSAA back buffer + a 493×570 4× MSAA back buffer + buffers).

With the workaround below (flush `_deferredDestroys` and call `wgpu.destroy()` right after `app.destroy()`) the growth drops to ~30 MB per round-trip and that remainder is collected by GC within ~60 s:

| round-trips | `graphics` before | `graphics` with workaround |
|---|---|---|
| 0 | 372 MB | 96 MB |
| 5 | 880 MB | 292 MB |
| 10 | 1,434 MB | 432 MB |
| idle 60 s | no change | 231 MB |

### Minimal reproduction

```js
import * as pc from 'playcanvas';

const canvas = document.createElement('canvas');
document.body.appendChild(canvas);
const device = await pc.createGraphicsDevice(canvas, { deviceTypes: [pc.DEVICETYPE_WEBGPU] });

const app = new pc.AppBase(canvas);
app.init({ graphicsDevice: device, componentSystems: [pc.RenderComponentSystem, pc.CameraComponentSystem], resourceHandlers: [] });
app.start();

// Any GPU resource: a 2048² texture is 16 MB, easy to see in a memory tool.
const texture = new pc.Texture(device, { width: 2048, height: 2048, format: pc.PIXELFORMAT_RGBA8, mipmaps: false });
texture.lock().fill(255); texture.unlock();
// Render a frame so the texture is actually uploaded.
app.render();
texture.destroy();          // -> device.deferDestroy(gpuTexture)

app.destroy();              // -> WebgpuGraphicsDevice.destroy()

console.log(device._deferredDestroys.length); // > 0: never flushed
console.log(device.wgpu);                     // still a live GPUDevice; GPUDevice.destroy() never called
```

Repeat "create device → render → destroy" in a loop and watch the process' GPU memory (Chrome task manager "GPU memory", or `footprint`/Activity Monitor on macOS): it only goes up.

### Expected behaviour

`WebgpuGraphicsDevice.destroy()` should

1. flush `_deferredDestroys` after `super.destroy()` (the resources have been queued by then; no `submit()` will ever run again on this device), and
2. call `this.wgpu.destroy()`. `handleDeviceLost` already treats `info.reason === 'destroyed'` (and `_destroyed`) as "do not recover", so this should not trigger the device-loss recovery path.

### Current workaround (application side)

```js
app.destroy();
for (const r of device._deferredDestroys) { try { r.destroy(); } catch {} }
device._deferredDestroys.length = 0;
device.wgpu?.destroy();
```

This touches private fields, so it will break whenever they are renamed — hence this issue.

### Environment

- PlayCanvas 2.22.0 (the same code is present on `main` as of 7e3ceef5 — `destroy()` → `destroyDeviceResources()` → `deferDestroy()` with no flush, and `wgpu.destroy()` only in `debugLoseContext()`)
- WKWebView (Tauri 2) on macOS 26, Apple Silicon. The deferred-destroy queue is backend logic, so the behaviour is the same on Chrome; only how aggressively the browser reclaims an unreferenced `GPUDevice` differs.

### PR DESCRIPTION
**Pull request by mvaligursky**
Fix GPU resource retention when repeatedly creating and destroying WebGPU applications, without waiting for browser garbage collection.

**Changes:**
- Release pending GPU resources after device teardown listeners finish, then destroy the native GPU device.
- Release resources immediately when asynchronous cleanup completes after device destruction.
- Add regression coverage for teardown ordering, normal deferred destruction, and pending readback cleanup.

Fixes #9393.

Thanks @tanitaka-tech for the detailed report, reproduction, memory measurements, and workaround!

### CODE REVIEW

File: src/platform/graphics/webgpu/webgpu-graphics-device.js [hunk 1]
```diff
@@ -340,6 +340,10 @@ class WebgpuGraphicsDevice extends GraphicsDevice {
         this.externalBackbuffer = null;
 
         super.destroy();
+
+        // Destroy listeners can enqueue more resources, and no further submit will drain them.
+        this.destroyDeferredResources();
+        this.wgpu?.destroy();
```
Comment (mvaligursky) [inline src/platform/graphics/webgpu/webgpu-graphics-device.js:374]: `GPUCanvasContext#unconfigure` is never called — it appears nowhere in `src/`, while `configure(this.canvasConfig)` runs on every `createDevice`. Per spec, `unconfigure()` "destroys any textures produced while configured", so it is the deterministic way to release the swapchain; without it those textures live until the canvas and its context are collected. For the canvases in the issue's measurements (a 1280x720 4x MSAA back buffer and a 493x570 one) that is several MB per canvas, multiply-buffered, per destroyed device — exactly the shape of the remainder that still needs GC. The configured context also keeps a reference to the now-destroyed `GPUDevice` through `canvasConfig.device`.

Adding `this.gpuContext?.unconfigure()` next to `this.wgpu?.destroy()` would cover it.

File: src/platform/graphics/webgpu/webgpu-graphics-device.js [hunk 2]
```diff
@@ -340,6 +340,10 @@ class WebgpuGraphicsDevice extends GraphicsDevice {
         this.externalBackbuffer = null;
 
         super.destroy();
+
+        // Destroy listeners can enqueue more resources, and no further submit will drain them.
```
Comment (mvaligursky) [inline src/platform/graphics/webgpu/webgpu-graphics-device.js:364]: Related: `destroy()` is less thorough than `loseContext()`, which starts by clearing exactly the state a dead device should not keep. Probing a destroyed device (real `destroy()`, stand-in GPU objects) shows what survives:

```
wgpu.destroy() called              true     <- the fix, working
_deferredDestroys.length           0        <- the fix, working
gpuContext.unconfigure() called    false
this.wgpu / gpuAdapter / gpuContext  still set
renderPipeline.cache.size          2
computePipeline.cache.size         1
commandEncoder still set           true
commandBuffers.length              1
bindGroupFormats.length            1
pipeline still set                 true
_bindGroups.size                   1
_bindGroupFormats.size             1
```

`loseContext()` clears `commandEncoder`, `commandBuffers`, `passEncoder`, `pipeline`, `bindGroupFormats` and both pipeline caches; `destroy()` clears none of them. This matters more than it looks: `GPURenderPipeline`, `GPUBindGroupLayout` and `GPUShaderModule` have no `destroy()` method, so dropping the JS reference is the *only* way to release them promptly — for a scene with many materials the cached pipelines are not a small allocation, and they are GC-only by construction. Reusing the same clearing block (or calling into a shared helper) would make `destroy()` at least as thorough as a device loss.

**Review by mvaligursky — COMMENTED** [review]
🤖 **Automated PR review** — posted on my behalf by Claude Code (Opus 4.8). Not a human review. The points below are suggestions to weigh as possible improvements, not changes that necessarily need to be addressed.

The ordering is right: `destroyDeviceResources()` runs before `super.destroy()` while `_destroyed` is still false, `GraphicsDevice#destroy` fires `'destroy'` before setting the flag so listener-enqueued resources land in the queue too, and the drain happens after both. `handleDeviceLost` already bails on `reason === 'destroyed'` and on `_destroyed`, so the `lost` promise `wgpu.destroy()` settles does not trigger a recreate. `GPUDevice#destroy` is idempotent, so the unguarded second `destroy()` is safe. 68 WebGPU tests pass, lint clean.

Both points below are about the residual the issue measures — "the growth drops to ~30 MB per round-trip and that remainder is collected by GC within ~60 s" — since the PR description claims cleanup "without waiting for browser garbage collection".

File: test/platform/graphics/webgpu/webgpu-context-restore.test.mjs [hunk 3]
```diff
@@ -49,6 +49,8 @@ describe('WebGPU context restoration', function () {
         device.renderPipeline = { cache: new Map() };
         device.computePipeline = new WebgpuComputePipeline(device);
         device.destroyDeviceResources = () => {};
+        device.destroyDeferredResources = WebgpuGraphicsDevice.prototype.destroyDeferredResources;
+        device.clearDeviceState = WebgpuGraphicsDevice.prototype.clearDeviceState;
```
Comment (mvaligursky) [inline test/platform/graphics/webgpu/webgpu-context-restore.test.mjs:53]: Minor: this is the second commit in a row that has to hand-copy a newly extracted private helper onto this fake device (`destroyDeferredResources` last commit, `clearDeviceState` this one), because the fake is a `NullGraphicsDevice` with ~18 WebGPU fields and methods grafted on and then driven via `WebgpuGraphicsDevice.prototype.loseContext.call(device, ...)`. It fails loudly rather than silently when a helper is missed, so this is friction rather than risk — but every future refactor of `loseContext` needs a matching edit here, and what is under test is a null device wearing WebGPU parts.

`webgpu-destroy.test.mjs` in this same PR shows the cheaper shape: `new WebgpuGraphicsDevice({ width: 1, height: 1 })` gives a real instance with its field initializers (which is exactly why `clearDeviceState` works there without wiring), and only `wgpu` / `gpuContext` need stubbing. Building this fake the same way would drop the per-helper wiring lines.

**Review by mvaligursky — COMMENTED** [review]
🤖 **Automated PR review** — posted on my behalf by Claude Code (Opus 4.8). Not a human review. The points below are suggestions to weigh as possible improvements, not changes that necessarily need to be addressed.

Both earlier points are closed. Probing a destroyed device (real `destroy()`, stand-in GPU objects), including a double `destroy()`:

```
gpuContext.unconfigure() calls   1
_deferredDestroys.length         0
this.wgpu / gpuAdapter / gpuContext / canvasConfig   all null
renderPipeline.cache.size        0
computePipeline.cache.size       0
commandEncoder                   null
commandBuffers.length            0
_bindGroups/_bindGroupFormats    0/0
_computes/_drawCommands          0/0
```

`clearDeviceState` is safe on a device that never initialized because `renderPipeline` / `computePipeline` are class field initializers, so the pre-init destroy test holds. The `gpuContext?.unconfigure()` guard matches the existing `gpuContext?.configure(...)` pattern. `_clusters.clear()` landed in `destroy()` rather than a per-frame reset, so cluster reuse is unaffected — and `WorldClustersAllocator#destroy` had no caller at all on `main`, so the lights and cluster textures of every allocated `WorldClusters` were leaking on app teardown independently of the WebGPU issue. 736 platform + scene tests pass, lint clean.

Nothing new of substance; one test-maintenance note below.
